docs/RFCS/20190906_flow_control.md

```diff
@@ -0,0 +1,137 @@
+- Latency: equals to the transaction executed latency, can be collected in DM-worker itself.
+- Bandwidth: this is a little complicated, but we can estimate it by adjusting the concurrency and find the maximum downstream tps for this DM-worker, which can be used as the bandwidth of the data import/replication link.
+
+Take the [benchmark result in DM 1.0-GA](https://pingcap.com/docs/v3.0/benchmark/dm-v1.0-ga/#benchmark-result-with-different-pool-size-in-load-unit) as an example, we extract the result in load unit test, use load unit pool size as X-axis, latency and load speed as Y-axis.
+
+![DM benchmark result with different pool size in load unit](../media/rfc-load-benchmark.png)
+
+From the benchmark result we can draw the following conclusions:
+
+- The latency grows exponentially as the concurrency increases.
+- The tps grows linely as the concurrency increases in a specific range, in this case the liner range is between 0-16.
+- After the tps reaches the maximum value, it remains the same even the concurrency increases. The maximum value of tps can be used as the bandwidth.
+
+#### How to detect and measure congestion
+
+Congestion can be exposed in different ways and have different effect on key indices. We have the following predictive indices change that may reflect congestion.
+
+- If tps doesn't increase alongside with the concurrency increase (assuming the dumped data or relay log stream is fast enough), we can believe the congestion is occurring.
+- If concurrency doesn't change, but tps decreases and latency increases significantly (maybe tps halves and latency doubles), we can believe the congestion is occurring.
+- If concurrency doesn't change, tps and latency don't change too much, but data loss rate increases significantly (however in most cases if data loss rate increases, we often find tps decreases and latency increases), we can believe the congestion is occurring.
+
+The congestion degree can be measured by the three indices, but the workflow in the real world is often complicated, we won't introduce an estimate model in this proposal, as we only need to known whether the congestion is occurring in our congestion control model.
+
+### Congestion control framework
+
+The congestion control framework is separated into three parts, the congestion detect component, the data flow control component and the concurrency control component.
+
+#### The congestion detect component
+
+The congestion detect component will collect the key indices in DM-worker as following:
+
+- data loss rate: we will add a sliding window to record the count of downstream resource busy error (such as `tmysql.ErrPDServerTimeout`, `tmysql.ErrTiKVServerTimeout`, `tmysql.ErrTiKVServerBusy` etc) and the count of the distributed transactions. The data loss rate will be calculated with these two counters.
```

Comment:
What does `distributed` mean in `the count of the distributed transactions`?

Reply:
typo here, changed to `dispatched`